openraft/src/raft/impl_raft_blocking_write.rs

```diff
@@ -30,17 +30,19 @@ where C: RaftTypeConfig<Responder = OneshotResponder<C>>
     /// - It proposes a **joint** config.
     /// - When the **joint** config is committed, it proposes a uniform config.
     ///
-    /// If `retain` is true, then all the members which not exists in the new membership,
+    /// If `retain` is `true`, then all the members which not exists in the new membership,
```

Comment:
Coming from #1191, I realized I am also confused about this `retain` argument:<

> If `retain` is `true`, then all the members which not exists in the new membership,

I kinda think it might be helpful to explicitly document that this `retain` argument is for removed voters, "all the members which not exists", IMHO this is ambiguous, a learner can no longer exist after a member change, so this statement is somewhat appliable to it as well,  though "converting a learner into a learner" does not make sense.

Maybe we can try something like:

```md
retain: after a membership change, if an old voter is removed from the voter list, 
the `retain` argument decides if it will remain in the cluster. If `true`, it will 
stay as a **learner**. Otherwise, it will be evicted from the cluster.
```


I checked the [implementation](https://github.com/datafuselabs/openraft/blob/ec284cb4813019d8b3a43a8cd74f3e1aa39eb882/openraft/src/membership/membership.rs#L282-L323), looks like this `retain` argument will be used for `ChangeMembers` variants:

```rs
RemoveVoters
ReplaceAllVoters
```

Since they are the only variants that can be used to remove a voter?

Reply:
> looks like this retain argument will be used for ChangeMembers variants:

If this `retain` argument is only used in these 2 variants, what about removing the it from the function and adding it to the associated types, like:

```rs
    pub async fn change_membership(
        &self,
        members: impl Into<ChangeMembers<C>>
    ) -> Result<ClientWriteResponse<C>, RaftError<C, ClientWriteError<C>>> {

pub enum ChangeMembers<C>
where C: RaftTypeConfig
{
    AddVoterIds(BTreeSet<C::NodeId>),
    AddVoters(BTreeMap<C::NodeId, C::Node>),
    RemoveVoters {
        node_ids: BTreeSet<C::NodeId>,
        retain: bool
    },
    ReplaceAllVoters {
        node_ids: BTreeSet<C::NodeId>,
        retain: bool
    },
    AddNodes(BTreeMap<C::NodeId, C::Node>),
    SetNodes(BTreeMap<C::NodeId, C::Node>),
    RemoveNodes(BTreeSet<C::NodeId>),
    ReplaceAllNodes(BTreeMap<C::NodeId, C::Node>),
}
```

then questions like #1191 would never happen again:)

Reply:
This change might affect the API `Raft::change_membership(m: impl Into<ChangeMembership>, retain:bool)`.

Most used cases are just to update voters with a BTreeMap: `Raft::change_membership(btreemap!{1=>(),2=>()})`, AFAIK. 

The `retain` flag is not related to the kind of membership changes internally inside Openraft, either.

Reply:
Ah, sorry for missing that, I forgot that it is a generic argument:<